platform/linux-generic/odp_ipsec.c

```diff
@@ -1368,7 +1468,8 @@ static int ipsec_out_esp(odp_packet_t *pkt,
 
 	param->auth_range.offset = ipsec_offset;
 	param->auth_range.length = state->ip_tot_len -
-				  state->ip_hdr_len -
+				  state->ip_hdr_len +
+				  ipsec_get_seqh_len(ipsec_sa) -
 				  ipsec_sa->icv_len;
 	param->hash_result_offset = state->ip_offset +
```

Comment:
This make hash result overlap with the ESN field. The code in this function is missing 1) making room in the packet for the 4 high order bytes of ESN for ICV calculation and 2) adjusting hash result offset when ESN is in use. Also, ipsec_out_esp_post() should have code for removing the high order ESN bytes and shifting the ICV to its correct place.

Reply:
Let me clarify here what we are doing:
1. We are intentionally making use of the ICV bytes place for inserting the ESN higher 4B.
2. The hash result(calculated ICV) is made to overwrite the existing ESN inserted higher 4B.
3. This way we avoided creating extra room for ESN higher 4B.

And here for "ODP_AUTH_ALG_NULL" case the "ipsec_get_seqh_len( )" will always return zero bytes.

So it seems to me that the code block functions correctly.

Reply:
Yes, I can see what you are trying to do, but the spec for hash_result_offset says this: "If this offset lies within specified 'auth_range', implementation will mute this field before calculating the hash result." Thus your high part of ESN will get cleared before the hash gets calculated.

Now that I look at the openssl based crypto implementation in linux-gen ODP, it appears that the hash field is cleared only when checking the hash but not when generating it. That seems to be a violation of the API spec. ODP DPDK does clear the field in both cases.

Reply:
> Now that I look at the openssl based crypto implementation in linux-gen ODP, it appears that the hash field is cleared only when checking the hash but not when generating it. That seems to be a violation of the API spec. ODP DPDK does clear the field in both cases.

Okay, as discussed in todays call, I will change this as per the current API spec.